eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -1487,12 +1501,18 @@ logWithCallStack callStack entry = impLogL %= (<> stack <> line <> indent 2 entr
             ]
         ]
     prefix n = if n <= 0 then "" else indent (n - 1) "└"
-    stack = vsep [prefix n <> prettySrcLoc' loc | (n, (_, loc)) <- zip [0, 2 ..] (getCallStack callStack)]
+    stack =
+      vsep
+        [prefix n <> prettySrcLoc' loc | (n, (_, loc)) <- zip [0, 2 ..] . reverse $ getCallStack callStack]
 
 -- | Adds a Doc to the log, which is only shown if the test fails
 logDoc :: HasCallStack => Doc AnsiStyle -> ImpTestM era ()
 logDoc = logWithCallStack ?callStack
 
+-- | Adds a pretty-printable value to the log, which is only shown if the test fails
+logPretty :: (Pretty a, HasCallStack) => a -> ImpTestM era ()
+logPretty = logWithCallStack ?callStack . pretty
+
```

Comment:
Why do we need this? It is not used anywhere.

We don't use `Pretty` instances for any of our types. We rely on `ToExpr` instead, that's why I'd rather avoid having helpers for `Pretty`.
```suggestion
```

Reply:
Actually, we do use `Prettyprinter` in a lot of places (26 modules import it) and we use the `pretty` method of `Pretty` quite a bit in this module. In this module it's just a quick way to convert from `String` to `Doc`, so we could use `fromString` for that. However, I also wanted a way to log a `Text` and `pretty` is an easy way of converting a `Text` to a `Doc`. (It's also a lot more efficient than `fromString . unpack` because `Doc` uses `Text` internally. `fromString` is implemented as `pretty . pack`!)

I could monomorphise `logPretty` and rename it to `logText` if that would be more acceptable. The implementation would be the same.

One of the reasons for wanting to log a `Text` is that `pShow` from `pretty-simple` returns a `Text`, and I find the output of `pShow` much more readable than `toExpr` because it's coloured and more compact (eg it doesn't turn all infix operators to prefix ones). I'm not suggesting we switch over to using `pretty-simple`, but it's very handy to be able to pull it in for debugging. Also, I notice quite a few other projects are using `pretty-simple` in one way or another.

Reply:
> Actually, we do use Prettyprinter in a lot of place

I am aware that we use `Prettyprinter` a lot. I want to avoid using `Pretty` type class, not `prettyprinter` package. The goal is to have a consistent way of dealing with test output for ledger types and the decision was made a while back that it will `ToExpr`, not `Pretty` or `pretty-simple`. There are incredible benefits that the former has over the latter.

> I could monomorphise logPretty and rename it to logText if that would be more acceptable. The implementation would be the same.

Yes, that would be better. I understand that everyone has their own ways of debugging. So, if adding such function will be helpful for you, then I have no problem with adding it. But, I want to make two points:

1. Before you add a function that you like personally, I'd suggest ensuring that it could be useful for others, not just you. So, will this `logText` be useful for anyone? I highly doubt that, because we have very minimal usage of `Text` in ledger.
2. I do not want to derail the whole project from the path that we have setup for our selves. That is why I really do not want to add `logPretty`. We already have a whole collection of "Pretty" garbage in this module that we need to get rid of: https://github.com/IntersectMBO/cardano-ledger/blob/master/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/PrettyCore.hs
I cringe every time I have to look at that module, so hopefully you can understand why I am so bitter about `Pretty`

If you have some compelling argument in favor of another library that we could use instead of `tree-diff`, I'll be happy to discuss it with you. Until then, we are not using `Pretty` or `pretty-simple` for any of the ledger types.

> I also wanted a way to log a Text 

Why didn't you just use `logDoc . pretty`? Adding implementation for a whole new function and then wasting time on this discussion already took way more time and effort than you would save if you just used `logDoc . pretty . pShow` during debugging.

>  (It's also a lot more efficient than fromString . unpack because Doc uses Text internally. fromString is implemented as pretty . pack!)

FYI. We are discussing implementation of a test suite here, where minor optimizations of log output are irrelevant. I could care less if we save a few nanoseconds while printing log output for a failing test!

> I notice quite a few other IOG projects are using pretty-simple in one way or another.

Most of projects at IOG use `tasty` and in my opinion this is a terrible choice. So, this is definitely not a great argument. The most valuable part of `tree-diff` is that it can do diffing! So, pretty colors and slightly more compact output have little to no value to me when compared to ability to diff. That being said, it doesn't mean that we can't evaluate other options that could improve our experience. It's just a PR that adds couple of test cases is not a time or a place for such discussion.

Reply:
I've changed it to a monomorphic function taking `Text`. LMK if you're happy with this.

Reply:
I personally don't see it as a useful function, but if it makes you happy, I am not gonna stand in the way :wink:

Reply:
I've decided to take that function out in my next PR. I hadn't seen your previous message when I wrote my last message in this thread (didn't refresh the page) and I hadn't thought about lazy text. Much better just to use `logDoc . pretty` when needed.